src/infill/ZigzagConnectorProcessor.cpp

```diff
@@ -0,0 +1,145 @@
+        }
+    }
+
+    const bool should_add =
+        (is_this_connection_even && !is_this_endpiece && !should_skip_this_connection) // normal connections that should be added
+        || (this->use_endpieces && is_this_endpiece);  // end piece if it is enabled;
+
+    return should_add;
+}
+
+
+void ZigzagConnectorProcessor::registerScanlineSegmentIntersection(const Point& intersection, int scanline_index, int direction)
+{
+
+    if (this->is_first_connector)
+    {
+        // process as the first connector if we haven't found one yet
+        // this will be processed with the last remaining piece at the end (when the polygon finishes)
+        this->first_connector.push_back(intersection);
+        this->first_connector_index = scanline_index;
+        this->first_connector_direction = direction;
+        this->is_first_connector = false;
+    }
+    else
+    {
+        // add this connector if needed
+        if (this->shouldAddThisConnector(this->last_connector_index, scanline_index, direction))
+        {
+            for (unsigned int point_idx = 1; point_idx < this->current_connector.size(); ++point_idx)
+            {
+                addLine(this->current_connector[point_idx - 1], this->current_connector[point_idx]);
+            }
+            // only add the last line if:
+            //  - it is not an end piece, or
+            //  - it is an end piece and "connected end pieces" is enabled
+            const bool is_this_endpiece = scanline_index == this->last_connector_index;
+            if (!is_this_endpiece || (is_this_endpiece && this->connected_endpieces))
+            {
+                addLine(this->current_connector.back(), intersection);
+            }
+        }
+    }
+
+    // update state
+    this->current_connector.clear(); // we're starting a new (odd) zigzag connector, so clear the old one
+    this->current_connector.push_back(intersection);
+    this->last_connector_index = scanline_index;
+    this->last_connector_direction = direction;
+}
+
+
+void ZigzagConnectorProcessor::registerPolyFinished()
+{
+    int scanline_start_index = this->last_connector_index;
+    int scanline_end_index = this->first_connector_index;
+    int direction = scanline_end_index > scanline_start_index ? 1 : -1;
+    const bool is_endpiece = this->is_first_connector || (!this->is_first_connector && scanline_start_index == scanline_end_index);
+
+    // decides whether to add this zag according to the following rules
+    if ((is_endpiece && this->use_endpieces)
+        || (!is_endpiece && this->shouldAddThisConnector(scanline_start_index, scanline_end_index, direction)))
+    {
+        // for convenience, put every point in one vector
+        for (const Point& point : this->first_connector)
+        {
+            this->current_connector.push_back(point);
+        }
+        this->first_connector.clear();
+
+        for (unsigned int point_idx = 1; point_idx < this->current_connector.size() - 1; ++point_idx)
```

Comment:
Please move the below loop and if-statement into a new function and remove the code duplication with the same code in `registerScanlineSegmentIntersection`.

Make a new function something like `void addCurrentZagConnector()`